nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,177 @@
+import tempfile
+
+from pathlib import Path
+
+from machine import Machine, NixStartScript
+
+
+class VLan ():
+    def __init__(self, nr: int, tmp_dir: Path):
+        self.nr = nr
+        self.socket_dir = tmp_dir / f"vde{self.nr}.ctl"
+
+        self.process: Optional[subprocess.Popen] = None
+        self.pid: Optional[int] = None
+        self.fd: Optional[os.file] = None
+
+        # TODO: don't side-effect environment here
+        os.environ[f"QEMU_VDE_SOCKET_{self.nr}"] = self.socket_dir
+
+    def start(self):
+        pty_master, pty_slave = pty.openpty()
+
+        self.process = subprocess.Popen(
+            ["vde_switch", "-s", self.socket_dir, "--dirmode", "0700"],
+            stdin=pty_slave,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=False,
+        )
+        self.pid = self.process.pid
+        self.fd = os.fdopen(pty_master, "w")
+        self.fd.write("version\n")
+
+        # TODO: perl version checks if this can be read from
+        # an if not, dies. we could hang here forever. Fix it.
+        assert self.process.stdout is not None
+        self.process.stdout.readline()
+        if not (self.socket_dir / "ctl").exists():
+            raise Exception("cannot start vde_switch")
+
+    def release(self):
+        self.fd.close()
+        self.process.terminate()
+
+
+class Driver():
+    def __init__(
+        self,
+        logger,
+        vm_scripts,
+        keep_vm_state=False,
+        configure_python_repl=id,
+        machine_class=Machine,
+        vlan_class=VLan,
+    ):
+        """
+        Args:
+            - configure_python_repl: a function to configure ptpython.repl
+        """
+        self.log = logger
+        self.configure_python_repl = configure_python_repl
+
+        tmp_dir = Path(os.environ.get("TMPDIR", tempfile.gettempdir()))
+        tmp_dir.mkdir(mode=0o700, exist_ok=True)
+
+        self.vlans = [
+            self.log(f"creating VDE switch for network {nr}") and
+            vlan_class(nr, tmp_dir) for nr in
+            list(dict.fromkeys(os.environ.get("VLANS", "").split()))
+        ]
+
+        def cmd(scripts):
+            for s in scripts:
+                yield NixStartScript(s)
+
+        self.machines = [
+            self.log(f"creating VM '{cmd.machine_name}'") and
+            machine_class(
+                start_command=cmd,
+                keep_vm_state=keep_vm_state,
+                name=cmd.machine_name,
+                log_serial=logger.log_serial,
+                log_machinestate=logger.log_machinestate,
+                tmp_dir=tmp_dir,
+            ) for cmd in cmd(vm_scripts)
+        ]
+
+        @atexit.register
+        def clean_up() -> None:
+            with self.log.nested("cleaning up"):
+                for machine in self.machines:
+                    if not machine.release():
+                        continue
+                    self.log(f"killed {machine.name} (pid {machine.pid})")
```

Comment:
this is only a nitpick, but you don't even need the `continue` if you invert the condition and put the log into the true branch.

Reply:
I'll check I guess the original reason was line length but now with `f"'` it's shorter :-)